Fix bug in contains when looking up a string in a non-monotonic datet…

tjader opened the pull request:
- [ ] closes #13572
- [x] tests added / passed
  pandas.tests.indexes.test_datetimelike.TestDatetimeIndex:test_contains
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

…ime index and the object in question is first in the index.

gfyoung committed:
CLN: Remove the engine parameter in CSVFormatter and to_csv

closes #13419
xref #11274

tjader committed:
Fix bug in contains when looking up a string in a non-monotonic datetime index and the object in question is first in the index.

tjader committed:
Fix bug in contains when looking up a string in a non-monotonic datetime index and the object in question is first in the index.

tjader committed:
Fix bug in contains when looking up a string in a non-monotonic datetime index and the object in question is first in the index.

tjader committed:
Fix bug in contains when looking up a string in a non-monotonic datetime index and the object in question is first in the index.

tjader committed:
Fix bug in contains when looking up a string in a non-monotonic datetime index and the object in question is first in the index.

sinhrks commented on pandas/tests/indexes/test_datetimelike.py:
```diff
@@ -721,6 +721,17 @@ def test_fillna_datetime64(self):
                            dtype=object)
             self.assert_index_equal(idx.fillna('x'), exp)
 
+    def test_contains(self):
```
can u also add tests for ``TimedeltaIndex`` and ``PeriodIndex``? (It looks work)

jreback replied:
you need to just move this to the ``Datetimelike`` class which tests all datetimelike indexes.

jreback replied:
further let's add testing with ``Timestamp`` and ``datetime`` as well (and a case where things are not there). finally add in a ``NaT`` both in the contained index as well as a tests with ``in``

adneu committed:
BUG: Groupby.nth includes group key inconsistently #12839

closes #12839

Author: adneu <aneumann31@gmail.com>

Closes #13316 from adneu/12839 and squashes the following commits:

16f5cd3 [adneu] Name change
ac1851a [adneu] Added docstrings/comments, and new tests.
4d73cbf [adneu] Updated tests
9b75df4 [adneu] BUG: Groupby.nth includes group key inconsistently #12839

TomAugspurger commented on pandas/tests/indexes/test_datetimelike.py:
```diff
@@ -721,6 +721,17 @@ def test_fillna_datetime64(self):
+        #GH13572
+        dates = ['2015-01-03', '2015-01-01', '2015-01-04', '2015-01-05', '2015-01-02']
+        monotonic = pd.to_datetime(sorted(dates))
+        non_monotonic = pd.to_datetime(['2015-01-03', '2015-01-01', '2015-01-04', '2015-01-05', '2015-01-02'])
```
You can reuse `dates` here in the to_datetime call.

tjader replied:
How is NaT supposed to work with period indices?
The only NaT-like object I can successfully test with is this one
```python
>>> pd.Period('NaT', freq='D') in pd.period_range('2015-01-01', periods=5, freq='D').insert(0, pd.NaT)
True
```
Any of `pd.NaT, None, float('nan'), np.nan` produce `False` as result, although they work on Datetimeindex objects with NaT's in them.

sinhrks replied:
Thx to confirm. Pls leave other nat-likes (``pd.NaT, None, float('nan'), np.nan``) ATM because ``PeriodIndex`` is not fully supported them yet (I'll cover in #12759, but PR is appreciated).

tjader replied:
Hmm, TimedeltaIndices have a funny behavior as well. Checking for NaT-like objects always returns true
```python
>>> None in pd.to_timedelta(range(5), unit='d') + pd.offsets.Hour(1)
True
```
Looks like `pandas.index.TimedeltaEngine.get_loc` doesn't handle NaT at all. It's probably possible to patch it on a higher level, but instinctively I feel that is where it should be handled.
@TomAugspurger / @jreback any thoughts?

tjader replied:
@sinhrks Thanks for clarifying. I've added #13582 for you.

tjader replied:
I've taken out the timedelta issues as well, and created a new PR #13603.

parthea committed:
In gbq, use googleapiclient instead of apiclient #13454 (#13458)

closes #13454

jreback committed:
COMPAT: 32-bit compat fixes mainly in testing

closes #13566
closes #13584

jorisvandenbossche committed:
RLS: switch master from 0.18.2 to 0.19.0 (#13586)

haleemur committed:
BUG: Datetime64Formatter not respecting ``formatter``

 - [x] closes #10690   - [x] tests added / passed   - [x] passes ``git
diff upstream/master | flake8 --diff``   - [x] whatsnew entry    the
Datetime64Formatter class did not accept a `formatter` argument, so
custom formatters passed in through `df.to_string` or `df.to_html`
were silently ignored.

Author: Haleemur Ali <haleemur@gmail.com>

This patch had conflicts when merged, resolved by
Committer: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Closes #13567 from haleemur/fix/dt64_outputformat and squashes the following commits:

8d84283 [Haleemur Ali] fix bug in Datetime64Formatter, which affected custom date formatted output for df.to_string, df.to_html methods

yui-knk committed:
BUG: Fix TimeDelta to Timedelta (#13600)

tjader committed:
Added more exhaustive tests for __contains__.

tjader committed:
Added more exhaustive tests for __contains__.

tjader committed:
Added more exhaustive tests for __contains__.

tjader committed:
Added more exhaustive tests for __contains__.

sinhrks committed:
BUG: DatetimeIndex - Period shows ununderstandable error

closes #13078

Author: sinhrks <sinhrks@gmail.com>

Closes #13581 from sinhrks/dti_period_error and squashes the following commits:

c957541 [sinhrks] BUG: DatetimeIndex - Period shows ununderstandable error

gfyoung committed:
ENH: add downcast to pd.to_numeric

Title is self-explanatory.  Closes #13352.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13425 from gfyoung/to-numeric-enhance and squashes the following commits:

4758dcc [gfyoung] ENH: add 'downcast' to pd.to_numeric

sinhrks committed:
CLN: remove radd workaround in ops.py

Remove workaround added in #353.

Author: sinhrks <sinhrks@gmail.com>

Closes #13606 from sinhrks/ops_radd_cln and squashes the following commits:

d873aad [sinhrks] CLN: remove radd workaround

sinhrks committed:
DEPR: rename Timestamp.offset to .freq

closes #12160

Author: sinhrks <sinhrks@gmail.com>

Closes #13593 from sinhrks/depr_timestamp_offset and squashes the following commits:

c7749d5 [sinhrks] DEPR: rename Timestamp.offset to .freq

sinhrks committed:
BUG: Block/DTI doesnt handle tzlocal properly

Author: sinhrks <sinhrks@gmail.com>

Closes #13583 from sinhrks/tzlocal and squashes the following commits:

93f59a3 [sinhrks] BUG: DTI doesnt handle tzlocal properly

sinhrks committed:
BUG: Series contains NaT with object dtype comparison incorrect (#13592)

closes #9005

sinhrks committed:
CLN/TST: Add tests for nan/nat mixed input (#13477)

closes #13467

jorisvandenbossche committed:
BUG: groupby apply on selected columns yielding scalar (GH13568) (#13585)

closes #13568

IamJeffG committed:
TST: Clean up tests of DataFrame.sort_{index,values} (#13496)

* TST: Clean up tests of DataFrame.sort_{index,values}

* Factor out Series sorting tests to own file.

* Delegate deprecated sort() and order() to their own tests.

Before this commit, the `Series.sort_values()` tests relied on deprecated
`Series.sort()` and `Series.order()` as the source of truth. However
they both merely called `Series.sort_values()` under the hood.

This commit consolidates the core test logic against `.sort_values()`
directly, while `.sort()` and `.order()` merely check for equivalence
with `.sort_values()`.

Also removes some no-op assertions that had rotted from the old days of
`sort()`/`order()`.

* Remove 'by' docstring from Series.sort_values

* Document defaults for optional sorting args

* Move more sort_values, sort_index tests to be together.

* Add test for Series.sort_index(sort_remaining=True)

* Improve `sort_values` tests when multiple `by`s

Duplicates values in the test DataFrame are necessary
to fully test this feature.

* PEP8 cleanup

* Annotate tests with GH issue

* Fix indentation - docstring string replacement

jorisvandenbossche committed:
DOC: asfreq clarify original NaNs are not filled (GH9963) (#13617)

sinhrks committed:
BUG: Invalid Timedelta op may raise ValueError

Author: sinhrks <sinhrks@gmail.com>

Closes #13624 from sinhrks/timedelta_comp and squashes the following commits:

856df95 [sinhrks] BUG: Invalid Timedelta op may raise ValueError

sinhrks committed:
CLN: Cleanup ops.py

Author: sinhrks <sinhrks@gmail.com>

Closes #13605 from sinhrks/ops_cln2 and squashes the following commits:

729997b [sinhrks] CLN: Cleanup ops.py

gfyoung committed:
CLN: Removed outtype in DataFrame.to_dict (#13627)

Follows up from #8486 in 0.15.0 by removing outtype in DataFrame.to_dict()

yui-knk committed:
CLN: Fix compile time warnings

This commit suppresses these warnings

warning: comparison of constant -1 with expression\
of type 'PANDAS_DATETIMEUNIT' is always true\
[-Wtautological-constant-out-of-range-compare]

Author: yui-knk <spiketeika@gmail.com>

Closes #13607 from yui-knk/fix_c_warning and squashes the following commits:

e9eee1d [yui-knk] CLN: Fix compile time warnings

jorisvandenbossche committed:
Pin IPython for doc build to 4.x (see #13639)

jreback committed:
CLN: reorg type inference & introspection

closes #12503

Author: Jeff Reback <jeff@reback.net>

Closes #13147 from jreback/types and squashes the following commits:

244649a [Jeff Reback] CLN: reorg type inference & introspection

gfyoung committed:
BLD: included pandas.api.* in setup.py (#13640)

jorisvandenbossche committed:
DOC/BLD: pin IPython version to 4.2.0 (#13639) (#13647)

sinhrks committed:
TST: reorganize tools.tests (#13619)

yarikoptic committed:
BF(TST): allow AttributeError being raised (in addition to TypeError) from mpl (#13641)

Closes #13570

gfyoung committed:
API, DEPR: Raise and Deprecate Reshape for Pandas Objects

Author: gfyoung <gfyoung17@gmail.com>

Closes #13012 from gfyoung/categorical-reshape-validate and squashes the following commits:

3ad161d [gfyoung] API: Prevent invalid arguments to Categorical.reshape

yui-knk committed:
CLN: Fix compile time warnings

Author: yui-knk <spiketeika@gmail.com>

Closes #13643 from yui-knk/warning2 and squashes the following commits:

ee3a4fb [yui-knk] CLN: Fix compile time warnings

jorisvandenbossche committed:
CLN: fix some issues in asv benchmark suite (#13630)

* CLN: fix params list

* Fix issue in asv.conf.json for win32+other environment

Fix mistaken exclusion of virtualenv or existing:same on win32 in the config.

Credits: @pv

* CLN: remove DataMatrix

* ASV: fix exlusion of tables package for non-conda environments

sinhrks committed:
TST: add tests for Timestamp.toordinal/fromordinal

follow-up for #13593

Author: sinhrks <sinhrks@gmail.com>

Closes #13610 from sinhrks/depr_timestamp_offset2 and squashes the following commits:

28f8d41 [sinhrks] TST: add tests for Timestamp.toordinal

toobaz committed:
CLN: Initialization coincides with mapping, hence with uniqueness check

 - [x] tests added / passed   - [x] passes ``git diff upstream/master
| flake8 --diff``    Rebased version of
https://github.com/pydata/pandas/pull/10229 which was [actually not](h
ttps://github.com/pydata/pandas/pull/10229#issuecomment-131470116)
fixed by https://github.com/pydata/pandas/pull/10199.    Nothing
particular relevant, just wanted to delete this branch locally and
noticed it still applies: you'll judge what to do of it.

Author: Pietro Battiston <me@pietrobattiston.it>

Closes #13594 from toobaz/fix_checkunique and squashes the following commits:

a63bd12 [Pietro Battiston] CLN: Initialization coincides with mapping, hence with uniqueness check

sinhrks committed:
API: Change Period('NAT') to return NaT

closes #12759
closes #13582

Author: sinhrks <sinhrks@gmail.com>

Closes #13609 from sinhrks/period_nat and squashes the following commits:

9305c36 [sinhrks] COMPAT: Period(NaT) now returns pd.NaT

jreback committed:
BUG: construction of Series with integers on windows not default to int64

closes #13646

Author: Jeff Reback <jeff@reback.net>

Closes #13661 from jreback/foo and squashes the following commits:

e26f9bf [Jeff Reback] BUG: construction of Series with integers on windows not defaulting to int64

gfyoung committed:
CLN: Removed levels attribute from Categorical

Deprecated back in `0.15.0` and therefore long overdue.  Closes #8376.

Author: gfyoung <gfyoung17@gmail.com>

Closes #13612 from gfyoung/categorical-levels-remove and squashes the following commits:

f1254df [gfyoung] MAINT: Relocated backwards compat categorical pickle tests
f3321cb [gfyoung] CLN: Removed levels attribute from Categorical

jreback commented:
needs a rebase. some of these tests may now be passing. #13609 fixed lots of things.

tjader committed:
Merge remote-tracking branch 'origin/bugfixes' into bugfixes

tjader commented:
Hmm, I think I made a mistake in rebasing. Looks like I pulled in every single change in to this pull request. I'll close this one and make a new one.
